x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,180 @@
 package keeper
 
 import (
-	// "fmt"
+	"fmt"
+	"strings"
 	"time"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	// gammtypes "github.com/osmosis-labs/osmosis/v15/x/gamm/types"
+
+	cltypes "github.com/osmosis-labs/osmosis/v15/x/concentrated-liquidity/types"
+	gammtypes "github.com/osmosis-labs/osmosis/v15/x/gamm/types"
+	"github.com/osmosis-labs/osmosis/v15/x/superfluid/types"
 )
 
-// UnlockAndMigrate unlocks a balancer pool lock, exits the pool and migrates the LP position to a full range concentrated liquidity position.
-// If the lock is superfluid delegated, it will undelegate the superfluid position.
+// MigrateLockedPositionFromBalancerToConcentrated unlocks a balancer pool lock, exits the pool and migrates the LP position to a full range concentrated liquidity position.
+// If the lock is superfluid delegated, it will undelegate the superfluid position and redelegate it as the concentrated liquidity position.
+// If the lock is superfluid undelegating, it will undelegate the superfluid position and redelegate it as the concentrated liquidity position, but continue to unlock where it left off.
```

Comment:
```suggestion
// If the lock is superfluid undelegating, it will instantly undelegate the superfluid position and redelegate it as the concentrated liquidity position, but continue to unlock where it left off.
```